flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.operators.sink.committables;
+
+import org.apache.flink.streaming.api.connector.sink2.CommittableSummary;
+import org.apache.flink.streaming.api.connector.sink2.CommittableWithLineage;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.OptionalLong;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Manages the committables coming from one subtask. */
+class SubtaskCommittableManager<CommT> {
+    private CommittableSummary<CommT> summary;
```

Comment:
Could we directly use a `int numCommittables` or `int numTotal` to bookkeep the total number of the committables? 

The only usage for other fields is the `drainCommitted` method below. I'm a bit concern on the right `subtaskId` to fill in here: if we want to keep the same protocol for the post-committer topology, the subtask id here should be the id of the current commit operator subtask? (Assume users might create operators in pre-committer topology that ends with a non-forward connection, or there is rescale of committer operator after restarting). Thus the subtaskId should be updated, and they could be passed from the `CheckpointManager`, which could also avoid replicate information between different level of state class~

Reply:
Good catch. I removed the committable summary from the `SubtaskCommittableManager` and only pass the necessary information now.